### ISSUE TITLE
OS customizations missing

### ISSUE THREAD

**Issue by rodizio1** [RespawnDespair/wifibroadcast-image-builder#57]
I think there are still some customizations missing (unfortunately, I never really made a list of changes I made, so this is from my memory):

- Systemd: Disable logging daemon and periodic compression of logs
- Systemd: Disable everything else it might do
- Clean-Up crontabs (we don't want cron to run any tasks at all)
- Make sure no other service/daemon is running other than "our" programs (i.e. wbc related stuff)
- Remove  packages that interfere with wifibroadcast function (wireless-regdb, crda, wpa_supplicant, and probably others I forgot right now)
- Remove packages that don't interfere with wifibroadcast function but are otherwise not necessary (for example spelling dictionary packages, popularity contest, etc. etc.)

**Comment by RespawnDespair** [comment]
Very true,
There are also a lot of kernel options being compiled now that are unnecessary.
If possible we need to make an inventory of everything that can be safely removed and i will add this to the scripts.
Perhaps you could provide some of the commands to achieve this.
I'll leave this issue open to aggregate such a list when we come upon them.
It's hard to find all modifications to the RC6 script...

**Comment by RespawnDespair** [comment]
I have removed two packages in the package script.
Every removal i do needs separate testing to see if i broke the image, so it will take some time, but work on this has started.

**Comment by RespawnDespair** [comment]
@rodizio1 
> * Systemd: Disable logging daemon and periodic compression of logs
> * Systemd: Disable everything else it might do
> * Clean-Up crontabs (we don't want cron to run any tasks at all)

Do you have the commands available for this? I'm not sure about the best way to approach this. (Not a long time Linux user, so some things still elude me).

**Comment by pilotnbr1** [comment]
Did you figure this out? If not I can look into it

**Comment by pilotnbr1** [comment]
Ive been working on it. The services mentioned are pretty easy to disable and I believe that disabling cron.service will kill all of its tasks. 

Just want to do some testing and also compare what is running on the beta image vs 1.6.

While making that comparison I can see what packages are installed on 1.6 so that should give us a pretty good idea of what to eliminate off the beta image on the first go without too much tedious code changes/rebuild cycles.

**Comment by pilotnbr1** [comment]
running systemctl status between 1.6 and 2.0beta so feel pretty good about killing these

#disable unneeded services
sudo systemctl disable cron.service
sudo systemctl disable syslog.service
sudo systemctl disable journald.service
sudo systemctl disable logind.service
sudo systemctl disable triggerhappy.service
sudo systemctl disable avahi-daemon.service
sudo systemctl disable ser2net.service
sudo systemctl disable dbus.service
sudo systemctl disable systemd-timesyncd.service

one reveal running sstemctl status is that the new image is running in a degraded mode which I think means some things failed to load. Upon investigation they are

hciuart.service (feel pretty good about killing this as it is for bluetooth support and currently wbc does not offer any- maybe in future)
AND
run.rpc_pipes.mount

The last does not mean anything to me- you guys? One way or another it should be addressed

**Comment by RespawnDespair** [comment]
That's some good investigative work, the hciuart could be because of the config.txt in boot not enabling bluetooth. 
The other one is unkown to me, we could disable it since it isn't starting anyway, but i'd feel better if i knew what it did. Google only gives me some hits about NFS mounts etc...
Since we have a mounted FS i don't think it is completely broken, but if somebody knows what this actually does, please chime in.
I'll add the other commands to a chroot script in the 04 WFB stage.

**Comment by pilotnbr1** [comment]
For what its worth: I was adding them in stage 4, 03-run-chroot.sh after the #disable sync option for usbmount. I did not test it yet with disabling hciuart.service but should will.

BTW
On 1.6 the only 2 services running not directly related to WBC (as far as I can tell) is:
udevd.service (might be a slight name change on 3b+ not totally sure)
ssh.service 

If we continue to have weird issues I may take a closer look at the timing order of how everything is started between 1.6 and 2.0. At a glance the order in which some of the WBC stuff start looks different but I cant say with certainty yet.

**Comment by RespawnDespair** [comment]
Once you are satisfied, make a pull request, i'll happily approve it.

**Comment by pilotnbr1** [comment]
Got side tracked by that run.rpc_pipes.mount error. Dug a bit in the boot logs and there is a failure to load module autofs4 . I believe this auto mounts partitions. This also explains some errors related to ext4 file system during boot. 

Going to compare all of this to the boot logs of 1.6 ANNNDD then finally get to the packages lol

**Comment by pilotnbr1** [comment]
Journald is still running. After some researching it is harder to kill than just “disable” in systemctl. 

I seem to vaguely remember this issue back in the day.. @rodizio1 any input?

**Comment by rodizio1** [comment]
Yeah, systemd is a very annoying piece of software. I don't remember what I did back then, but I remember at some point I started deleting stuff because I couldn't find another way to cleanly disable this systemd crap.

One option would be using the tool "meld" to compare directories and files of my image with the new image (thing is just, since the image-builder uses a newer Raspbian based on a newer Debian version, things could have changed significantly).

**Comment by pilotnbr1** [comment]
yes I am seeing that.. It is whack a mole with systemd. When I delete packages to kill one service sometimes it comes back in another form. This is going to take more work. I am still plugging away

**Comment by pilotnbr1** [comment]
Closing in on some solutions.

The workflow is time consuming, especially when modifying the build early on. 
I have been creating an image then running it and making mods to the live image at command line. Finally going back and implementing this changes in the build script and testing. It can be slow.. as you know. It’s also tempting to try and make too many changes in one build run and thus complicating debugging lol

So I should have a pull for purges packages, stopped services, and later v4l2loopback support although having more issues than expected getting it to make

**Comment by RespawnDespair** [comment]
I know, the power of the image builder is more apparent when changing in the later steps. The kernel build for instance took a good long while to get right.
Making modifications in the image, writing them down and then doing several at once is how i do it myself.
Should the original pull request still be done? Or are you going to replace it? v4l would be great to get working!

**Comment by pilotnbr1** [comment]
No please wait on the pull, the original pull needs to be modified. I am waiting for an entire build to cook and then test. 

I think solution for for that journald is using a mask command. Essentially break it buy sending it’s output to the infamous null file. This may be the solution for some other service along with purging of packages. @rodizio1 does this sound familiar

My understanding of systemd is it is more powerful than init.d in that if a new service has a service dependency it will start new services. Everything tries to write to journald thus systemd is restarting it constantly.

**Comment by Yes21** [comment]
> I know, the power of the image builder is more apparent when changing in the later steps. The kernel build for instance took a good long while to get right.

One other big benefit of the building script is that every one can see and reproduce how the image has been built.
That's not the case for all others image (except DroneBridge) where there's no documentation about their building process.
So @RespawnDespair, I say you again many thanks for this great tool :-)

**Comment by pilotnbr1** [comment]
Yes +1 @Yes21 !

The pull request for the services killer commit is ready to go. 

Their is a new pull request for package deletion. 

Both were briefly tested on pi 3b and 3b+ acting as groundpi. Video and telemetry OK. They need more testing as AirPi and the RC part.

**Comment by pilotnbr1** [comment]
So far these are the packages I removed and tested in my pull request

# Remove packages that conflict with the workings of EZ-Wifibroadcast
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge wireless-regdb
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge wpasupplicant
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge crda
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge cron

#Untested Packages for possible removal
#DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge dbus
#DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge dhcpcd5
#DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge isc-dhcp-client
#DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge isc-dhcp-common


# Remove packages for space savings
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge apt-transport-https
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge aptitude
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge aptitude-common
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge apt-listchanges
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge avahi-daemon
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge bzip2
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge cifs_utils
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge curl
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge iptables
DEBIAN_FRONTEND=noninteractive sudo apt-get -yq purge triggerhappy

Does anything jump out that might need to stay? I have come to the conclusion that I could spend hours and hours just getting rid of unneeded packages... The 4 commented packages are ones that might be purged as well that involve networking. Thoughts?

Rodizio had a good idea with using meld and doing a comparison. Might look into that later but I kind of see the space savings stuff as something we could get to later (lower priority). Additionally there are other space saving that could be implimented such as removing all info files and more...

**Comment by Yes21** [comment]
@pilotnbr1 Many thanks for all thi great job :-)
I will test it as far as your PRs will be merged.

**Comment by pilotnbr1** [comment]
@Yes21 if you want to test you should remove the skip files from stage 3 and stage 4

**Comment by Yes21** [comment]
Thanks, I know it. I've already rerun stages 3 and 4 many times today !!

**Comment by Yes21** [comment]
Could we close this issue now ?

**Comment by RespawnDespair** [comment]
@pilotnbr1 are you happy with your changes? Any loose ends?
I'm confident the kernel config part is okay now.
If you are happy we'll close this issue as well...

**Comment by pilotnbr1** [comment]
Its looking good!

**Comment by RespawnDespair** [comment]
I agree, great work!
